pypeman/remoteadmin.py

```diff
@@ -0,0 +1,342 @@
+
+@contextlib.contextmanager
+def stdoutIO(stdout=None):
+    old = sys.stdout
+    if stdout is None:
+        stdout = StringIO()
+    sys.stdout = stdout
+    yield stdout
+    sys.stdout = old
+
+class RemoteAdminServer():
+    """
+    Expose json/rpc function to a client by a websocket.
+    """
+
+    def __init__(self, host, port):
+        self.host = host
+        self.port = port
+        self.ctx = {}
+
+    def get_channel(self, name):
+        """
+        return channel by is name.all
+        """
+        for chan in channels.all:
+            if chan.name == name:
+                return chan
+        return None
+
+    async def start(self):
+        """ Start remote admin server """
+        methods.add(self.exec)
+        methods.add(self.channels)
+        methods.add(self.stop_channel)
+        methods.add(self.start_channel)
+        methods.add(self.list_msg)
+        methods.add(self.replay_msg)
+        methods.add(self.push_msg)
+
+        start_server = websockets.serve(
+            self.command,
+            self.host,
+            self.port
+        )
+        await start_server
+
+    async def command(self, websocket, path):
+        """
+        Generic function to handle a command from client.
+        """
+        request = await websocket.recv()
+        response = await methods.dispatch(request)
+        if not response.is_notification:
+            await websocket.send(str(response))
+
+    async def exec(self, command):
+        """
+        Execute a python command on this instance and
+        return the stdout result.
+
+        :param command: The python command to execute. Can be multiline.
+        :returns: Command stdout result.
+        """
+        # TODO may cause problem on multi thread access to stdout
+        # as we highjack sys.stdout
+        with stdoutIO() as out:
+            exec(command, globals())
+
+        return out.getvalue()
+
+    async def channels(self):
+        """
+        Return a list of available channels.
+        """
+        chans = []
+        for chan in channels.all:
+            if not chan.parent:
+                chans.append({
+                    'name': chan.name,
+                    'status': channels.BaseChannel.status_id_to_str(chan.status)
+                })
+        return chans
+
+    async def start_channel(self, channel):
+        """
+        Start the specified channel
+
+        :params channel: The channel name to start.
+        """
+        chan = self.get_channel(channel)
+        await chan.start()
+
+    async def stop_channel(self, channel):
+        """
+        Stop the specified channel
+
+        :params channel: The channel name to stop.
+        """
+        chan = self.get_channel(channel)
+        await chan.start()
+
+    async def list_msg(self, channel):
+        """
+        List first 10 messages from message store of specified channel.
+
+        :params channel: The channel name.
+        """
+        # TODO allow indexing
```

Comment:
why just 10 ? param maybe

Reply:
Yes, I want to do it in a next PR as I have to modify message store also...